src/internal/connector/exchange/api/mail_test.go

```diff
@@ -145,3 +153,202 @@ func (suite *MailAPIUnitSuite) TestMailInfo() {
 		})
 	}
 }
+
+type MailAPIE2ESuite struct {
+	tester.Suite
+	credentials account.M365Config
+	ac          Client
+	user        string
+}
+
+// We do end up mocking the actual request, but creating the rest
+// similar to E2E suite
+func TestMailAPIE2ESuite(t *testing.T) {
+	suite.Run(t, &MailAPIE2ESuite{
+		Suite: tester.NewIntegrationSuite(
+			t,
+			[][]string{tester.M365AcctCredEnvs},
+		),
+	})
+}
+
+func (suite *MailAPIE2ESuite) SetupSuite() {
+	t := suite.T()
+
+	a := tester.NewM365Account(t)
+	m365, err := a.M365Config()
+	require.NoError(t, err, clues.ToCore(err))
+
+	suite.credentials = m365
+	suite.ac, err = NewMockableClient(m365)
+	require.NoError(t, err, clues.ToCore(err))
+
+	suite.user = tester.M365UserID(suite.T())
+}
+
+func getJsonObject(t *testing.T, thing serialization.Parsable) map[string]interface{} {
+	sw := kjson.NewJsonSerializationWriter()
+
+	err := sw.WriteObjectValue("", thing)
+	require.NoError(t, err, "serialize")
+
+	content, err := sw.GetSerializedContent()
+	require.NoError(t, err, "serialize")
+
+	var out map[string]interface{}
+	err = json.Unmarshal([]byte(content), &out)
+	require.NoError(t, err, "unmarshall")
+
+	return out
+}
+
+func (suite *MailAPIE2ESuite) TestHugeAttachmentListDownload() {
+	mid := "fake-message-id"
+	aid := "fake-attachment-id"
+
+	tests := []struct {
+		name            string
+		setupf          func()
+		attachmentCount int
+		expect          assert.ErrorAssertionFunc
+	}{
+		{
+			name: "simple fetch",
+			setupf: func() {
+				mitem := models.NewMessage()
+				mitem.SetId(&mid)
+
+				gock.New("https://graph.microsoft.com").
```

Comment:
I was exploring how to test this and using something like `gock` seemed like a good choice. Let me know if you have any thoughts or other options.

I could think of 3 options:
- Option 1: Use [golang/mock](https://github.com/golang/mock) to create mocks for SDK  
	- ❌ Have to run codegen
	- ✅ Can run tests in parallel
- Option 2: Use [h2non/gock](https://github.com/h2non/gock) to intercept and substitute http requests  
	- ✅ Don't need a codegen  
	- ✅ We don't have to make any assumptions about SDK
	- ❌ Not thread safe and cannot run parallely  
- Inject a middleware to respond with custom http response  
	- ❌ Have to create implementation from scratch  
	- ✅ Don't have to make assumptions about SDK behaviour apart from middlewares  


/cc @ashmrtn @ryanfkeepers

Reply:
While I'm not usually a fan of external mocking packages (largely because they're normally overly heavy and come with their own corner cases), where a lightweight, local interface will do just as well.  But I don't think we're going to do any better in this case by writing anything locally.  Let's keep with Gock here, and we can revisit as needed.

Other notes:

> ❌ Have to run codegen

This is a big no for me.  Having to run codegen for mocking is absolutely something to avoid.

> Not thread safe and cannot run parallely

Only a problem if we register overlapping urls, right?  That can be avoided by standardizing around random uuids/strings as test vars instead of hardcoded ids.

Reply:
> Only a problem if we register overlapping urls, right?

Yup, AFAICT if we have separate urls, we should good.

> I love seeing tests like this as it gives us so much more control. ...

Yeah, I really like how gock fits it, especially the fact that we don't have to make assumptions about how the SDK works. This also has the advantage of making sure any SDK changes also gets tested. The setup is also relatively minimal for `gock`.